### PR TITLE
Winter Holiday Update! (pt 2)

### PR DESCRIPTION
**Pull request by youkitan**
- Overhaul linearization pipeline
- Improved datatype organization

### CODE REVIEW

File: code-matlab/shared/datatypes/tc/tc.m [hunk 1]
```diff
@@ -1,84 +1,50 @@
 function tc_out = tc(varargin)
 %% TC TUNING CURVE DATATYPE
+%   The tc dataype is a container for tuning curves typically used in experimental
+%   analysis.
+%
 %   tc_out = TC(varargin) is a constructor for the tc (tuning curve) struct.
 %
-% youkitan 2014-10-27
+%   INPUTS:
+%
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: add description here

also, may be useful to point to an example workflow that uses this data type -- is there one?

File: code-matlab/shared/datatypes/tc/tc.m [hunk 2]
```diff
@@ -1,84 +1,50 @@
...
+%
+%   OUTPUTS
+%       iv_out: iv struct with fields:
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: iv_out? should be tc_out right?

File: code-matlab/shared/datatypes/tc/tc.m [hunk 3]
```diff
@@ -1,84 +1,50 @@
...
+%          .tc - [NxM] array, where N is the number of cells and M is the number of
+%                tuning variable bins. Each entry is a normalized firing rate.
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: why a *normalized* firing rate? if you simply mean spike counts divided by occupancy, that's a *firing rate*. Normalized suggests something like, normalized suich that 1 is the max firing rate over some stimulus dimension, which is not what we want I think.

File: code-matlab/shared/datatypes/tc/tc.m (hunk 2 above)
Comment (youkitan) [inline code-matlab/shared/datatypes/tc/tc.m]: fixed in upcoming commit

File: code-matlab/shared/datatypes/tc/tc.m [hunk 4]
```diff
@@ -1,84 +1,50 @@
...
+% youkitan edit Feb 2017 overhaul
 
 %% Set fields
 tc_out.tc = []; %nCells x nBins
-tc_out.occ = []; %
-tc_out.label = {};
-tc_out.units = {};
+tc_out.occ_hist = []; %
+tc_out.spk_hist = [];
+tc_out.binEdges = [];
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: these aren't described in the output section of the help

File: code-matlab/shared/datatypes/tc/tc.m (hunk 1 above)
Comment (youkitan) [inline code-matlab/shared/datatypes/tc/tc.m]: - in terms of input, I was going to use the tsd constructor as a template
- in terms of examples, there are workflows in code-matlab/example_workflows but I haven't updated those (they use the "old" versions of equivalent functions/datatype)

File: code-matlab/shared/datatypes/tc/tc.m [hunk 5]
```diff
@@ -1,84 +1,50 @@
...
-        tc_out.tc = varargin{3};
+        tc_out.occ_hist = varargin{2};
+        tc_out.spk_hist = varargin{3};
         tc_out.tc = varargin{4};
 end
 
-% Enforce equal sized arrays
-for i = 1:length(tc_out.tc)
-    % Check to make sure vectors are of same size
-    assert(length(tc_out.firing_rates{i}) == length(tc_out.stimulus{1,i}),...
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: what's this `stimulus` field? doesn't appear to be defined earlier?

also, `out.firing_rates` isn't described in the help

File: code-matlab/shared/datatypes/tc/tc.m [hunk 6]
```diff
@@ -1,84 +1,50 @@
...
-   fprintf('The stimulus for cell %d has %d number of dimensions',i,dims(1));
-   
-end %iterate cells
-
-
-
-%% Set defaults
-tc_out.firing_rates = {}; %nCells x nBins
-tc_out.stimulus = {}; %nCells x nStimulus
-tc_out.parameters = [];
```
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: I'm going to stop commenting here other than the general point to make the help consistent with what's happening here

File: code-matlab/shared/proc/TuningCurves.m [hunk 7]
```diff
@@ -188,25 +191,44 @@
...
+if exist('tc2D','var'); tc_out.tc2D = tc2D; end
+tc_out.occ_hist = occ_hist;
+tc_out.spk_hist = spk_hist;
+tc_out.binEdges = binEdges;
+tc_out.binCenters = binCenters;
+tc_out.nBins = nBins;
 
+% usr data
+tc_out.usr.good_idx = good_idx;
+tc_out.usr.no_occ_idx = good_idx;
```
Comment (mvdm) [inline code-matlab/shared/proc/TuningCurves.m]: `good_idx` goes to both of these? doesn't seem right

also, could describe this in help

File: code-matlab/shared/datatypes/tc/tc.m (hunk 1 above)
Comment (mvdm) [inline code-matlab/shared/datatypes/tc/tc.m]: 1) OK
2) right, so if this new `tc` isn't backwards compatible (is it?) should create an issue

File: code-matlab/shared/linearize/LinearizePos.m [hunk 8]
```diff
@@ -1,78 +1,154 @@
...
-% OUTPUTS:
+%   CFG OPTIONS:
+%       cfg_def.debugMode = 0; calculate distance from pos_tsd to coord
+%       cfg_def.verbose = 1;
+%       cfg_def.nanMethod = 1; % default 1, use nans; if 2, remove nans
+%       cfg_def.outputType = 'idx'; If 'idx', returns index of coord points. If 'dist',
+%       returns distance in units of run_dist.
 %
-% linpos_tsd: tsd with linearized position (z field) and distance (z_dist)
+%   LINEARIZATION EXAMPLE:
```
Comment (mvdm) [inline code-matlab/shared/linearize/LinearizePos.m]: nice!

File: code-matlab/shared/linearize/StandardizeCoord.m [hunk 9]
```diff
@@ -1,34 +1,87 @@
...
-Coord_out(1,:) = interp1(1:size(Coord_in,2),Coord_in(1,:),linspace(1,size(Coord_in,2),nBins),'linear');
-Coord_out(2,:) = interp1(1:size(Coord_in,2),Coord_in(2,:),linspace(1,size(Coord_in,2),nBins),'linear');
+%% hosuekeeping
+Coord_out.coord = coord_std;
+Coord_out.run_dist = run_dist;
+Coord_out.nPoints = nPoints;
+Coord_out.pointDist = pointDist;
+Coord_out.standardized = 1;
 
+end
```
Comment (mvdm) [inline code-matlab/shared/linearize/StandardizeCoord.m]: overall I really like this, it's a big improvement over the previous haphazard situation!

an example usage would be great here too (can create issue if not yet available), perhaps even with illustrations/tutorial style instructions along the lines of what you showed me in your office earlier. this part of the analysis workflow has the potential to be quite confusing yet it's the first thing that happens in many analyses!

File: code-matlab/shared/datatypes/tc/tc.m (hunk 5 above)
Comment (youkitan) [inline code-matlab/shared/datatypes/tc/tc.m]: these no longer exist (was an old version of the tc datatype)

File: code-matlab/shared/datatypes/tc/tc.m (hunk 6 above)
Comment (youkitan) [inline code-matlab/shared/datatypes/tc/tc.m]: same as above

**Pull request merged**